packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -0,0 +1,181 @@
+        return BaseRelayRecipient._msgData();
+    }
+
+    constructor(
+        IERC20 _weth,
+        IERC20 _token,
+//        IQuoter _quoter,
+        IRelayHub _relayHub,
+        ISwapRouter _uniswap,
+        IChainlinkOracle _priceFeed,
+        address _trustedForwarder,
+        uint24 _uniswapPoolFee,
+        uint256 _gasUsedByPost,
+        uint256 _permitMethodDataLength,
+        string memory _permitMethodSignature
+    ) {
+        weth = _weth;
+        token = _token;
+//        quoter = _quoter;
+        uniswap = _uniswap;
+        priceFeed = _priceFeed;
+        uniswapPoolFee = _uniswapPoolFee;
+        permitMethodSignature = bytes4(keccak256(bytes(_permitMethodSignature)));
+        permitMethodDataLength = _permitMethodDataLength;
+        setRelayHub(_relayHub);
+        setPostGasUsage(_gasUsedByPost);
+        setTrustedForwarder(_trustedForwarder);
+        // allow uniswap to transfer from paymaster balance
+        token.approve(address(uniswap), uint256(-1));
+    }
+
+    /**
+     * set gas used by postRelayedCall, for proper gas calculation.
+     * You can use TokenGasCalculator to calculate these values (they depend on actual code of postRelayedCall,
+     * but also the gas usage of the token and of Uniswap)
+     */
+    function setPostGasUsage(uint256 _gasUsedByPost) public onlyOwner {
+        gasUsedByPost = _gasUsedByPost;
+    }
+
+    function _calculatePreCharge(
+        GsnTypes.RelayRequest calldata relayRequest,
+        uint256 maxPossibleGas)
+    internal
+    view
+    returns (uint256 tokenPreCharge) {
+        uint256 ethMaxCharge =
+            relayHub.calculateCharge(maxPossibleGas, relayRequest.relayData) +
+            relayRequest.request.value;
+        uint256 price = uint256(priceFeed.latestAnswer());
+        uint256 decimals = uint256(priceFeed.decimals());
+        tokenPreCharge = ethMaxCharge.mul(price).div(10**decimals);
+    }
+
+    function preRelayedCall(
+        GsnTypes.RelayRequest calldata relayRequest,
+        bytes calldata signature,
+        bytes calldata approvalData,
+        uint256 maxPossibleGas
+    )
+    external
+    override
+    relayHubOnly
+    returns (bytes memory context, bool revertOnRecipientRevert) {
+        (signature);
+        require(approvalData.length == 0, "approvalData: invalid length");
+        if (relayRequest.relayData.paymasterData.length != 0) {
+            require(
+                permitMethodSignature == GsnUtils.getMethodSig(relayRequest.relayData.paymasterData),
+                "paymasterData: wrong method sig");
+            // execute permit method for this token
+            // solhint-disable-next-line avoid-low-level-calls
+            (bool success, bytes memory ret) = address(token).call(relayRequest.relayData.paymasterData);
+            require(success, string(abi.encodePacked("permit call reverted:", string(ret))));
```

Comment:
nitpic: this revert string is not really a string...its either Error(String) or Panic(uint)
(but don't have much else to do, so maybe garbled string is the best way...)

Reply:
do we decode these Error/Panic arrays to strings anywhere? At least this is a cheap way to propagate this error message to the UI, and yes, it won't look like a string there but it will be decodable. What do you suggest I do with it?

Reply:
No, no better way, as it can be any error format. That's why it's only a nitpic: just annoyed, but nothing to do about it.